crates/ty_python_semantic/resources/mdtest/intersection_types.md

```diff
@@ -317,35 +324,49 @@ def _(
     i4: Intersection[Not[P], Q, P],
     i5: Intersection[P, Any, Not[P]],
     i6: Intersection[Not[P], Any, P],
+    i7: Intersection[R[P], Not[R[P]]],
+    i8: Intersection[R[P], Not[R[Q]]],
 ) -> None:
     reveal_type(i1)  # revealed: Never
     reveal_type(i2)  # revealed: Never
     reveal_type(i3)  # revealed: Never
     reveal_type(i4)  # revealed: Never
     reveal_type(i5)  # revealed: Never
     reveal_type(i6)  # revealed: Never
+    reveal_type(i7)  # revealed: Never
+    reveal_type(i8)  # revealed: R[P] & ~R[Q]
 ```
 
 ### Union of a type and its negation
 
 Similarly, if we have both `P` and `~P` in a _union_, we can simplify that to `object`.
 
+```toml
+[environment]
+python-version = "3.12"
+```
+
 ```py
 from ty_extensions import Intersection, Not
 
 class P: ...
 class Q: ...
+class R[T]: ...
 
 def _(
     i1: P | Not[P],
     i2: Not[P] | P,
     i3: P | Q | Not[P],
     i4: Not[P] | Q | P,
+    i5: R[P] | Not[R[P]],
+    i6: R[P] | Not[R[Q]],
 ) -> None:
     reveal_type(i1)  # revealed: object
     reveal_type(i2)  # revealed: object
     reveal_type(i3)  # revealed: object
     reveal_type(i4)  # revealed: object
+    reveal_type(i5)  # revealed: object
+    reveal_type(i6)  # revealed: R[P] | ~R[Q]
```

Comment:
Similar to above, this depends on variance. If bivariant (as the current definition of `R` would be) this is `object`; if invariant it should simplify to `~R[Q]`, if covariant or contravariant it should stay unsimplified.